Rocket.Chat/Controllers/Chat/MessagesViewControllerComposerDelegate.swift

```diff
@@ -24,27 +25,102 @@ extension MessagesViewController: ComposerViewExpandedDelegate {
     // MARK: Hints
 
     func composerView(_ composerView: ComposerView, didChangeHintPrefixedWord word: String) {
-        return
+        viewModel.hints = []
+        viewModel.hintPrefixedWord = word
+
+        guard
+            let realm = Realm.current,
+            let prefix = viewModel.hintPrefixedWord.first
+        else {
+            return
+        }
+
+        let word = String(word.dropFirst())
+
+        if prefix == "@" {
+            viewModel.hints = User.search(usernameContaining: word, preference: []).map { $0.0 }
+
+            if "here".contains(word) || word.count == 0 {
+                viewModel.hints.append("here")
+            }
+
+            if "all".contains(word) || word.count == 0 {
+                viewModel.hints.append("all")
+            }
+        } else if prefix == "#" {
+            let filter = "auth != nil && (privateType == 'c' || privateType == 'p')\(word.isEmpty ? "" : "&& name BEGINSWITH[c] %@")"
+
+            let channels = realm.objects(Subscription.self).filter(filter, word)
+
+            for channel in channels {
+                viewModel.hints.append(channel.name)
+            }
+
+        } else if prefix == "/" {
+            let commands: Results<Command>
+            if word.count > 0 {
+                commands = realm.objects(Command.self).filter("command BEGINSWITH[c] %@", word)
+            } else {
+                commands = realm.objects(Command.self)
+            }
+
+            commands.forEach {
+                viewModel.hints.append($0.command)
+            }
+        } else if prefix == ":" {
+            let emojis = EmojiSearcher.standard.search(shortname: word.lowercased(), custom: CustomEmoji.emojis())
+
+            emojis.forEach {
+                viewModel.hints.append($0.suggestion)
+            }
+        }
```

Comment:
What do you think about moving this method to the view model as well? Maybe we can have one view model just for the hints? 🤔

Reply:
I saw that you guys are doing this pattern, so I did the same. I created a MessageComposerViewModel and put this logic there :+1:

Reply:
👍